src/warc2zim/utils.py

```diff
@@ -42,14 +53,53 @@ def parse_title(content):
         return ""
 
 
-def to_string(input_: str | bytes) -> str:
-    try:
-        input_ = input_.decode(  # pyright: ignore[reportGeneralTypeIssues, reportAttributeAccessIssue]
-            "utf-8-sig"
-        )
-    except AttributeError:
-        pass
-    return input_  # pyright: ignore[reportGeneralTypeIssues, reportReturnType]
+def get_record_encoding(record: ArcWarcRecord) -> str | None:
+    content_type = get_record_content_type(record)
+    if m := ENCODING_RE.search(content_type):
+        return m.group("encoding")
+
+
+def to_string(input_: str | bytes, encoding: str | None) -> str:
+    tried_encodings = set()
+    if isinstance(input_, str):
+        return input_
+
+    if not input_:
+        # Empty bytes are easy to decode
+        return ""
+
+    if encoding:
+        try:
+            return input_.decode(encoding)
+        except (ValueError, LookupError):
+            tried_encodings.add(encoding)
+            pass
+
+    # Detect encoding from content.
+    # First try declared encoding in content.
```

Comment:
If you add a comment stating "First try declared encoding in content", you should add another comment stating "Then try ..."

And I would rather change this to the list you mentioned in your PR (first use the HTTP headers, then ..., then ...) with a link to HTML specification you mentioned.

This should probably be placed in the docstring of the method in fact.

Reply:
Added a docstring.